### PR TITLE
feat(pooling): load Sentence Transformers CrossEncoder heads

### LINKED ISSUE #54981: [Feature]: Support Sentence Transformers Transformer-Pooling-Dense CrossEncoders
### 🚀 The feature, motivation and pitch

Sentence Transformers can save CrossEncoders as explicit module chains. The standard layouts include:

- `Transformer` for sequence classification
- `Transformer → LogitScore` for causal-LM rerankers
- `Transformer → Pooling → Dense` for a feature-extraction backbone with a trained scoring head

The first layout already maps to vLLM's sequence-classification path, and vLLM has model-specific support for some causal-LM rerankers. The remaining gap is that vLLM does not interpret the saved Sentence Transformers module stack and its configs.

An immediate use case is a multimodal cross-encoder reranker saved as `Transformer → Pooling → Dense`. The Dense layer contains trained checkpoint weights, while the Pooling and Transformer configs define how hidden states and structured multimodal pairs are processed. Sentence Transformers can reload the complete model, but vLLM currently cannot discover the scoring head or reproduce those saved semantics without model-specific code or launch-time overrides.

This issue proposes bounded support for the standard serialized layout, starting with the exact `Transformer → Pooling → Dense` chain:

- read the CrossEncoder and module metadata saved with the checkpoint
- map supported pooling modes to vLLM's pooling runner
- load the trained Dense scoring head and its activation
- preserve supported structured multimodal input formatting
- fail clearly when the checkpoint uses unsupported modules or semantics
- leave non-pooled CrossEncoders on their existing vLLM paths

This would let the same checkpoint load in Sentence Transformers and vLLM without `trust_remote_code` or a model-specific vLLM implementation.

### Alternatives

- Add a model-specific sequence-classification implementation in Transformers. This makes the checkpoint portable to Transformers, but fixes the pooling and scoring design at the model-class boundary and requires upstream support for each model family.
- Add another model-specific reranker path in vLLM. This would solve one checkpoint but duplicate behavior already represented by the Sentence Transformers metadata.
- Require custom checkpoint code or manual vLLM flags. This works, but makes the checkpoint less portable and easier to load incorrectly.

### Additional context

- [Sentence Transformers custom CrossEncoder models](https://github.com/huggingface/sentence-transformers/blob/main/docs/cross_encoder/usage/custom_models.rst)
- [Earlier vLLM exploration of modular CrossEncoder metadata](https://github.com/vllm-project/vllm/pull/31563)
- [Prototype implementation](https://github.com/vllm-project/vllm/compare/main...oliverholworthy:oholworthy/feat/sentence-transformers-cross-encoder)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by oliverholworthy**
## Purpose

Closes #54981.

Sentence Transformers can serialize CrossEncoders as explicit module stacks, but vLLM does not currently reconstruct the Transformer → Pooling → Dense layout. The saved pooling behavior and trained Dense scoring head therefore require custom code or manual configuration.

This PR adds bounded support for that exact layout:

- maps cls, mean, and lasttoken pooling to vLLM
- loads the saved Dense weights, bias, and activation
- supports structured text and multimodal query-document messages using the saved chat template
- uses the explicit multimodal text configuration when selecting bidirectional attention
- rejects unsupported pooled configurations rather than silently changing their semantics
- leaves existing non-pooled sequence-classification and CausalLM plus LogitScore paths unchanged

## Test plan

```bash
python -m pytest -q \
  tests/entrypoints/pooling/scoring/test_sentence_transformers_io_processor.py \
  tests/models/test_adapters.py \
  tests/transformers_utils/test_config.py \
  tests/transformers_utils/test_repo_utils.py \
  tests/models/transformers/test_backend.py::test_attention_class_selection \
  -k 'cross_encoder or dense_loader or sequence_classification or sentence_transformers or attention_class_selection'
```

## Test result

- 34 passed, 29 deselected
- local pre-commit hooks passed, including Ruff, mypy, SPDX, and configuration validation
- loaded a custom-code-free Mistral 3 VL checkpoint derived from nvidia/rc1-nemotron-3.5-embed with trust_remote_code=False
- scored text, image, and image-plus-text pairs in Sentence Transformers and vLLM; maximum absolute difference was 0.0122

### CODE REVIEW

**Review by tomaarsen — COMMENTED** [review]
From the Sentence Transformers side, I’m supportive of this approach. Transformer -> Pooling -> Dense is a valid and very reasonable CrossEncoder layout that gives a lot of flexibility with e.g. multimodality in ST. Overall, I'm a big fan of the ST data being parsed and loaded directly, so both vLLM and ST work from the same checkpoint. I also think it's smart to read `modules.json` like you're doing and explicitly reject unsupported ST module configurations (i.e. that don't map to anything meaningful in vLLM at this time). It's also smart to check the `transformer_config` to make sure its a `feature-extraction` ('base model') variant.

I believe the other common variant, just a Transformer except it's a `sequence-classification` variant, is also still supported via the `config.json` where it recognizes the `transformers` classification head.

And then another common format, Transformer -> LogitScore (with the Transformer as a `text-generation` or `any-to-any` variant) is explicitly not implemented here yet, which is okay. That setup 'generates' the logits for the next token, and then either grabs the logits of a specific token ID (e.g. for "yes" or "1") or subtracts the logits of 2 specific token IDs (e.g. "yes" - "no" or "1" - "0") to produce a score.

Lastly, the reported multimodal parity check would probably work well as an integration test. The tests that save a model with Sentence Transformers, load it in vLLM, and compare the scores are a good setup in my opinion.

cc @hmellor as I know you've been interested in easier embedding model deployment as well. Sentence Transformers is really just transformers with some extra small layers on top (dense, pooling, normalize, etc.). It's pretty accessible for vLLM I think.

File: vllm/model_executor/models/transformers/base.py [hunk 1]
```diff
@@ -675,9 +683,14 @@ def _get_attn_cls(self) -> type[AttentionLayerBase]:
         is_encoder = lambda module: not getattr(module, "is_causal", True)
         has_encoder = lambda model: any(is_encoder(m) for m in model.modules())
         is_multimodal = lambda config: config != config.get_text_config()
+        text_is_encoder = getattr(self.text_config, "is_causal", None) is False
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/base.py:686]: In Transformers `config.is_causal` is extremely rare, only existing in 4 configs (3 of which are Gemma 4 variants).

Is this a standard field in configs coming from Sentence Transformers?

Comment (tomaarsen) [reply vllm/model_executor/models/transformers/base.py:686]: This is a relatively common "magic field" added in Transformers v5.2.0: https://github.com/huggingface/transformers/pull/43705
In short, decorators in `transformers` will pass the config `is_causal` to the attention module and mask creation. It turns the mask bidirectional, as you might've expected. Embedding models and rerankers get a pretty consistent 1-2% improvement with bidirectional attention at a low cost (after all, you can't reuse KV caches with embeddings/rerankers anyways), so I think the majority of new embedding/reranker models are bidirectional (probably still 30-40% of new models just stick with the original causal backbone though).

File: vllm/model_executor/models/transformers/legacy.py [hunk 2]
```diff
@@ -54,14 +58,51 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
...
+                "use enforce_eager=True."
+            )
 
     def forward(
         self,
         input_ids: torch.Tensor | None,
         positions: torch.Tensor,
         intermediate_tensors: IntermediateTensors | None = None,
         inputs_embeds: torch.Tensor | None = None,
+        token_type_ids: torch.Tensor | None = None,
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/legacy.py:91]: This should arrive in `kwargs` otherwise it's a required argument for any encoder-attention model. This actually already happens on `main`

File: vllm/model_executor/models/transformers/legacy.py [hunk 3]
```diff
@@ -54,14 +58,51 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
...
+            torch.zeros(
+                vllm_config.scheduler_config.max_num_batched_tokens,
+                dtype=torch.int32,
+                device=self.device_config.device,
+            )
+            if getattr(self.text_config, "type_vocab_size", 0) > 0
+            and "token_type_ids" in signature(self.model.forward).parameters
+            else None,
+            persistent=False,
+        )
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/legacy.py:72]: Why is it necessary to create our own buffer here? Can this not be handled by the `dynamic_arg_dims`?

File: vllm/model_executor/models/transformers/base.py (hunk 1 above)
Comment (hmellor) [reply vllm/model_executor/models/transformers/base.py:686]: I hadn't realised that this field came in via decorators, in that case this is probably fine.

Now that we have `AttentionFuser` which allows us to inspect the actual HF attention module, we should really make the attn cls detection lazy so we don't need to use these heuristics
